### ISSUE TITLE
Publishing messages when server is down results in lots of threads and connections

### ISSUE THREAD

**Issue by Yaytay** [vert-x3/vertx-rabbitmq-client#90]
I have a test that sends 1M messages to rabbit (as fast as it can receive HTTP requests).
If the test starts whilst the RabbitMQ server is down the client correctly builds up the requests and the messages do get sent through when the server comes up.
However, at this point the debugger is showing that there are lots (hundreds) of threads in the client process.

I have a secondary issue that if the server is taken down after the connection has been established it breaks the client and I haven't found a way around that.

Note that I'm working with vertx4, milestone 3.

**Comment by Yaytay** [comment]
Not quite sure of the vertx way to do this (is a syncronized block permitted), but I think that start() needs to work with a single Future per instance.
If start is called when the Future already exists it should just return the single Future, if the Future is currently null it should set it from the existing code.
?

**Comment by vietj** [comment]
@Yaytay you can use synchronized blocks, can you provide a patch to improve this ?

**Comment by Yaytay** [comment]
I've done a quick patch that just chains the Handlers together, but it's not robust (if one of the handlers fails it will lose things - minimally it'll lose whatever that handler was meant to do, but as I've written it right now it'll lose all the subsequent handlers, either way I'll lose messages unless there is something else).
So I'm starting to think that, for my purposes, the abstraction of vertx-rabbitmq-client is wrong.

I think that what I actually need is a vert.x Event Bus handler layered on top of vertx-rabbitmq-client that sends all messages that reach it on to Rabbit.  This would resolve my current issue of having hundreds of concurrent attempts to open a connection (at the expense of flooding the vertx bus).
I'll have a go at that and you can take a look to see whether you'd like it back in vertx.

**Comment by vietj** [comment]
which "Handlers" are you referring to ?

**Comment by Yaytay** [comment]
The AsyncResult handlers following a successful connection attempt.
My naive approach started like this:
`
  private void start(int attempts, Handler<AsyncResult<Void>> resultHandler) {
    synchronized(startLock) {
      if (startHandler != null) {
        Handler<AsyncResult<Void>> oldHandler = startHandler;
        startHandler = ar-> {
          oldHandler.handle(ar);
          resultHandler.handle(ar);
        };
      } else {
        startHandler = resultHandler;
      }
    }
    vertx.<Void>executeBlocking(future -> {
          try {
            connect();
            future.complete();
          } catch (IOException | TimeoutException e) {
            log.error("Could not connect to rabbitmq", e);
            future.fail(e);
          }
        }, ar -> {
          if (ar.succeeded() || retries == null) {
            Handler<AsyncResult<Void>> handler;
            synchronized(startLock) {
              handler = startHandler;
              startHandler = null;
            }
            handler.handle(ar);
          } else if (attempts >= retries) {
            log.info("Max number of connect attempts (" + retries + ") reached. Will not attempt to connect again");
            Handler<AsyncResult<Void>> handler;
            synchronized(startLock) {
              handler = startHandler;
              startHandler = null;
            }
            handler.handle(ar);
          } else {
            long delay = config.getConnectionRetryDelay();
            log.info("Attempting to reconnect to rabbitmq...");
            vertx.setTimer(delay, id -> {
              log.debug("Reconnect attempt # " + attempts);
              start(attempts + 1, resultHandler);
            });
          }
        });
  }
`
Basically daisy chaining those handlers, but each of those handlers will contain a call to basicPublish and I don't want to lose any of those calls.
That is difficult to manage when that handler does more than just call basicPublish - i.e. an exception in setting up the channel after establishing the connection would be hard to distinguish from an exception in the call to basicPublish itself - but in the former case I definitely don't want to lose the call to basicPublish.

Therefore I think I should just collect the message data from overlapping calls to publish and then process them once I have a suitably prepared channel.

**Comment by Yaytay** [comment]
I have been working on a synchronous wrapper for rabbit that (for publishing):
- queues up messages and sends them as quickly as it can
- survives restarting the server (this is the key bit for our reliability)
- resends unconfirmed messages after a reconnect.
- uses a generic message ID (not the delivery tag) to confirm sends to the caller without needing to wait (uses a ConfirmListener)
For consuming the handling is closer to what you've already got, but with:
- survives restarting the server 
- support for batch acking

Fitting this into the existing vertx-rabbitmq-client would mean:
- a few changes to RabbitMQClient (I know I need support for addConfirmListener, presumably converting into a ReadStream, there will probably be others).
- a new RabbitMQProducer 
- changes to RabbitMQConsumer to reset when there is an unexpected bad shutdown event
- changes to support batch acking, but I haven't looked into this yet

Are these changes that you would like in vertx-rabbitmq-client and, if so, would you like one massive PR or lots of smaller ones (that will probably have to be cumulative)?

**Comment by Yaytay** [comment]
Fixed by RabbitMQPublisher.
